sshd-core/src/main/java/org/apache/sshd/client/keyverifier/KnownHostsServerKeyVerifier.java

```diff
@@ -153,35 +153,43 @@ public void setModifiedServerKeyAcceptor(ModifiedServerKeyAcceptor acceptor) {
 
     @Override
     public boolean verifyServerKey(ClientSession clientSession, SocketAddress remoteAddress, PublicKey serverKey) {
-        Collection<HostEntryPair> knownHosts = getLoadedHostsEntries();
         try {
             if (checkReloadRequired()) {
                 Path file = getPath();
                 if (exists()) {
-                    knownHosts = reloadKnownHosts(clientSession, file);
+                    updateReloadAttributes();
+                    keysSupplier = GenericUtils.memoizeLock(getKnownHostSupplier(clientSession, file));
```

Comment:
Update of keysSupplier field is not mutually exclusive...

Reply:
Yes, but should not be an issue. If you have an idea how to make it even more robust I am happy to implement it..

Reply:
While this might not be 100% proof I prefer the following:

```java
private final AtomicReference<Supplier<? extends Collection<HostEntryPair>>> keysSupplierHolder = new AtomicReference<>(getKnownHostSupplier(null, getPath()));
```

And then replace the relevant sections with:
```java
keysSupplierHolder.set(....);
```